File: biosilico-creatures/src/main/resources/creatures/articlesAroundAndScripts/scripts/creatures1genomeAnalysis.pl
#!/usr/bin/perl
use strict;
use warnings;
use JSON::PP;

=head1 NAME
creatures1genomeAnalysis.pl - Parse and analyze Creatures genome files.

=head1 SYNOPSIS
perl creatures1genomeAnalysis.pl [filename.gen]

=head1 DESCRIPTION
This script parses Creatures genome files (`.gen`) and extracts gene information.
It supports all gene types and subtypes, including Brain Lobes, Biochemistry Reactions, and Creature Instincts.
The output includes human-readable descriptions and JSON exports.

=head1 GENE STRUCTURE
- Header: 8 bytes (gene marker, type, subtype, etc.)
- Data: Variable length, parsed based on gene type/subtype.

=head1 GENE TYPES AND SUBTYPES
=over 4
=item * Brain Genes (Type 0x00)
=over 8
=item - Lobe (Subtype 0x00): Defines brain lobes with position and size.
=back
=item * Biochemistry Genes (Type 0x01)
=over 8
=item - Receptor (Subtype 0x00): Detects chemical concentrations.
=item - Emitter (Subtype 0x01): Emits chemicals.
=item - Reaction (Subtype 0x02): Chemical reactions.
=item - Half-Lives (Subtype 0x03): Chemical decay rates.
=item - Initial Concentration (Subtype 0x04): Initial chemical levels.
=back
=item * Creature Genes (Type 0x02)
=over 8
=item - Stimulus (Subtype 0x00): Stimulus responses.
=item - Genus (Subtype 0x01): Species and parent information.
=item - Appearance (Subtype 0x02): Visual appearance.
=item - Instinct (Subtype 0x05): Innate behaviors.
=back

=cut

# --- Reference Data Sources ---
# Source: creatures1BrainMapCells_GenesHeader.txt
my %LOBE_NAMES = (
    0 => "Perception",
    1 => "Drive",
    2 => "Stimulus Source",
    3 => "Verb",
    4 => "Noun",
    5 => "General Sense",
    6 => "Decision",
    7 => "Attention",
    8 => "Concept",
    9 => "Regulator",  # Creatures 2+
);

# Source: creaturesDevelopmentRessources.pdf, p.8-12
my %CHEM_NAMES = (
    0x00 => "Pain",
    0x01 => "Need for Pleasure",
    0x02 => "Hunger",
    0x03 => "Coldness",
    0x04 => "Hotness",
    0x0A => "Fear",
    0x0B => "Boredom",
    0x1E => "Reward",
    0x1F => "Punishment",
    # Additional chemicals can be added here from the documentation.
);

# Source: creatures1BrainMapCells_GenesHeader.txt, p.21
my %ACTION_NAMES = (
    0 => "Quiescent",
    1 => "Push (Activate 1)",
    2 => "Pull (Activate 2)",
    3 => "Stop (Deactivate)",
    4 => "Come (Approach)",
    5 => "Run (Retreat)",
    6 => "Get",
    7 => "Drop",
    8 => "Think/Say",
    9 => "Sleep/Rest",
    10 => "Left",
    11 => "Right",
);

# --- Gene Class ---
package Gene;
use Data::Dumper;

=head2 Gene->new(%args)
Creates a new Gene object.

=cut
sub new {
    my ($class, %args) = @_;
    my $self = {
        # Required fields
        type        => defined $args{type}        ? $args{type}        : 0,
        subtype     => defined $args{subtype}     ? $args{subtype}     : 0,
        number      => defined $args{number}      ? $args{number}      : 0,
        switch_on   => defined $args{switch_on}   ? $args{switch_on}   : 0,
        # Optional fields
        sex_dep     => defined $args{sex_dep}     ? $args{sex_dep}     : "None",
        mutability  => defined $args{mutability}  ? $args{mutability}  : "None",
        data        => defined $args{data}        ? $args{data}        : [],
        parsed_data => defined $args{parsed_data} ? $args{parsed_data} : {},
    };
    bless $self, $class;
    return $self;
}

=head2 parse_reaction()
Parses Reaction genes (Biochemistry type).

=cut
sub parse_reaction {
    my ($self) = @_;
    return unless defined $self->{type} && $self->{type} == 0x01 &&
                  defined $self->{subtype} && $self->{subtype} == 0x02;
    return unless ref($self->{data}) eq 'ARRAY' && @{$self->{data}} >= 4;

    my @data = @{$self->{data}};
    my @reactants = ();
    my @products = ();
    for my $i (0..3) {
        last if $i + 1 >= @data;
        my $chem = $data[$i + 1] || 0;
        my $proportion = $data[$i] || 0;
        my $container = ($i < 2) ? \@reactants : \@products;
        push @$container, {
            chem => $chem,
            proportion => $proportion,
            chem_name => $CHEM_NAMES{$chem} || sprintf("Chem%02X", $chem),
        };
    }

    my $rate = (@data >= 9 && defined $data[8]) ? $data[8] : 1;
    $self->{parsed_data} = {
        reactants => \@reactants,
        products => \@products,
        rate => $rate,
    };
}

=head2 parse_instinct()
Parses Instinct genes (Creature type).

=cut
sub parse_instinct {
    my ($self) = @_;
    return unless defined $self->{type} && $self->{type} == 0x02 &&
                  defined $self->{subtype} && $self->{subtype} == 0x05;
    return unless ref($self->{data}) eq 'ARRAY' && @{$self->{data}} >= 9;

    my @data = @{$self->{data}};
    my ($lobe1, $cell1, $lobe2, $cell2, $lobe3, $cell3, $action, $reward_punish, $amount) = @data[0..8];
    $self->{parsed_data} = {
        conditions => [
            { lobe => $LOBE_NAMES{$lobe1} || sprintf("Lobe%02X", $lobe1 || 0), cell => $cell1 || 0 },
            { lobe => $LOBE_NAMES{$lobe2} || sprintf("Lobe%02X", $lobe2 || 0), cell => $cell2 || 0 },
            { lobe => $LOBE_NAMES{$lobe3} || sprintf("Lobe%02X", $lobe3 || 0), cell => $cell3 || 0 },
        ],
        action => $ACTION_NAMES{$action} || sprintf("Action%02X", $action || 0),
        reward_punish => (defined $reward_punish && $reward_punish == 0) ? "Reward" : "Punish",
        amount => $amount || 0,
    };
}

=head2 parse_lobe()
Parses Brain Lobe genes.

=cut
sub parse_lobe {
    my ($self) = @_;
    return unless defined $self->{type} && $self->{type} == 0x00 &&
                  defined $self->{subtype} && $self->{subtype} == 0x00;
    return unless ref($self->{data}) eq 'ARRAY' && @{$self->{data}} >= 5;

    my ($x, $y, $width, $height, $perception_link) = @{$self->{data}}[0..4];
    $self->{parsed_data} = {
        position => { x => $x || 0, y => $y || 0 },
        size => { width => $width || 0, height => $height || 0 },
        perception_link => defined $perception_link ? ($perception_link ? "Yes" : "No") : "No",
        neurons => ($width || 0) * ($height || 0),
    };
}

=head2 to_string()
Returns a human-readable string representation of the gene.

=cut
sub to_string {
    my ($self) = @_;

    my $str = sprintf(
        "Gene %03d: Type=0x%02X (%s), Subtype=0x%02X (%s), SwitchOn=%d (%s), SexDep=%s, Mutability=%s",
        $self->{number} || 0,
        $self->{type} || 0, Gene::_get_type_name($self->{type}),
        $self->{subtype} || 0, Gene::_get_subtype_name($self->{type}, $self->{subtype}),
        $self->{switch_on} || 0, Gene::_get_switch_on_name($self->{switch_on}),
        $self->{sex_dep} || "None",
        $self->{mutability} || "None",
    );

    # Add details for Reaction genes
    if (defined $self->{type} && $self->{type} == 0x01 &&
        defined $self->{subtype} && $self->{subtype} == 0x02 &&
        ref($self->{parsed_data}) eq 'HASH' && exists $self->{parsed_data}{reactants}) {
        my $reaction = $self->{parsed_data};
        $str .= "\n  Reaction: " .
            join(" + ", map { ($_->{proportion} || 0) . " " . ($_->{chem_name} || "Unknown") } @{$reaction->{reactants}}) . " → " .
            join(" + ", map { ($_->{proportion} || 0) . " " . ($_->{chem_name} || "Unknown") } @{$reaction->{products}}) .
            " (Rate: " . ($reaction->{rate} || 1) . ")";
    }
    # Add details for Instinct genes
    elsif (defined $self->{type} && $self->{type} == 0x02 &&
           defined $self->{subtype} && $self->{subtype} == 0x05 &&
           ref($self->{parsed_data}) eq 'HASH' && exists $self->{parsed_data}{conditions}) {
        my $instinct = $self->{parsed_data};
        $str .= "\n  Instinct: IF " .
            join(" AND ", map { ($_->{lobe} || "Unknown") . "[" . ($_->{cell} || 0) . "]" } @{$instinct->{conditions}}) .
            " THEN " . ($instinct->{action} || "Unknown") .
            " (" . ($instinct->{reward_punish} || "Unknown") . ": " . ($instinct->{amount} || 0) . ")";
    }
    # Add details for Brain Lobe genes
    elsif (defined $self->{type} && $self->{type} == 0x00 &&
           defined $self->{subtype} && $self->{subtype} == 0x00 &&
           ref($self->{parsed_data}) eq 'HASH' && exists $self->{parsed_data}{position}) {
        my $lobe = $self->{parsed_data};
        $str .= "\n  Lobe: " . ($LOBE_NAMES{$self->{number}} || sprintf("Lobe%02X", $self->{number} || 0)) .
            " at (" . ($lobe->{position}{x} || 0) . "," . ($lobe->{position}{y} || 0) . "), " .
            "Size: " . ($lobe->{size}{width} || 0) . "x" . ($lobe->{size}{height} || 0) .
            ", Neurons: " . ($lobe->{neurons} || 0) .
            ", Perception Link: " . ($lobe->{perception_link} || "No");
    }

    return $str;
}

# Helper functions
sub _get_type_name {
    my ($type) = @_;
    my %TYPE_NAMES = (
        0x00 => "Brain",
        0x01 => "Biochemistry",
        0x02 => "Creature",
    );
    return $TYPE_NAMES{$type} || "Unknown";
}

sub _get_subtype_name {
    my ($type, $subtype) = @_;
    my %SUBTYPE_NAMES = (
        0x00 => { 0x00 => "Lobe" },
        0x01 => {
            0x00 => "Receptor",
            0x01 => "Emitter",
            0x02 => "Reaction",
            0x03 => "Half-Lives",
            0x04 => "Initial Concentration",
        },
        0x02 => {
            0x00 => "Stimulus",
            0x01 => "Genus",
            0x02 => "Appearance",
            0x03 => "Pose",
            0x04 => "Gait",
            0x05 => "Instinct",
            0x06 => "Pigment",
            0x07 => "Pigment Bleed",
        },
    );
    return $SUBTYPE_NAMES{$type}{$subtype} || "Unknown";
}

sub _get_switch_on_name {
    my ($stage) = @_;
    my @STAGE_NAMES = qw(Embryo Child Youth Adolescent Adult Senior Old);
    return $STAGE_NAMES[$stage] || "Unknown";
}

package main;

=head1 FUNCTIONS

=head2 parse_genome_file($filename)
Parses a .gen file and returns a list of Gene objects.

=cut
sub parse_genome_file {
    my ($filename) = @_;
    open(my $fh, "<:raw", $filename) or die "Cannot open $filename: $!";
    binmode($fh);
    my $content = do { local $/; <$fh> };
    close($fh);

    my @genes;
    my $offset = 0;

    while ($offset < length($content)) {
        my $header = substr($content, $offset, 9);
        last unless length($header) >= 8;  # 8 bytes minimum for header

        my ($gene_marker, $type, $subtype, $number, $unknown, $switch_on) =
            unpack("A4 C C C C C", substr($header, 0, 8));

        my $sex_mut = (length($header) >= 9) ? unpack("C", substr($header, 8, 1)) : 0;
        my $data_length = (length($content) > $offset + 9) ? unpack("C", substr($content, $offset + 9, 1)) : 0;
        $data_length //= 0;

        $offset += 1;  # Skip invalid marker
        next unless $gene_marker eq "gene";

        # Parse sex dependency and mutability
        my $sex_dep = ($sex_mut & 0x10) ? "Female" : (($sex_mut & 0x08) ? "Male" : "None");
        my @mutability;
        push @mutability, "Mutable" if $sex_mut & 0x01;
        push @mutability, "Duplicable" if $sex_mut & 0x02;
        push @mutability, "Deletable" if $sex_mut & 0x04;
        my $mutability = @mutability ? join(",", @mutability) : "None";

        # Parse gene data
        my $data = [];
        if ($data_length > 0 && $offset + 10 + $data_length <= length($content)) {
            $data = [unpack("C*", substr($content, $offset + 10, $data_length))];
        }

        my $gene = Gene->new(
            type        => $type,
            subtype     => $subtype,
            number      => $number,
            switch_on   => $switch_on,
            sex_dep     => $sex_dep,
            mutability  => $mutability,
            data        => $data,
        );

        eval { $gene->parse_reaction() };
        eval { $gene->parse_instinct() };
        eval { $gene->parse_lobe() };

        push @genes, $gene;
        $offset += 10 + $data_length;
    }

    return \@genes;
}

=head2 export_to_json($genes, $filename)
Exports gene data to a JSON file.

=cut
sub export_to_json {
    my ($genes, $filename) = @_;
    my @json_genes;
    for my $gene (@$genes) {
        my $json_gene = {
            type        => sprintf("0x%02X", $gene->{type} || 0),
            type_name   => Gene::_get_type_name($gene->{type}),
            subtype     => sprintf("0x%02X", $gene->{subtype} || 0),
            subtype_name => Gene::_get_subtype_name($gene->{type}, $gene->{subtype}),
            number      => $gene->{number} || 0,
            switch_on   => $gene->{switch_on} || 0,
            switch_on_name => Gene::_get_switch_on_name($gene->{switch_on}),
            sex_dep     => $gene->{sex_dep} || "None",
            mutability  => $gene->{mutability} || "None",
            data        => $gene->{data},
        };
        $json_gene->{parsed_data} = $gene->{parsed_data} if ref($gene->{parsed_data}) eq 'HASH';
        push @json_genes, $json_gene;
    }
    open(my $json_fh, ">", $filename) or die "Cannot open $filename: $!";
    print $json_fh JSON::PP->new->pretty->encode(\@json_genes);
    close($json_fh);
}

=head2 display_genes($genes)
Displays gene information in a human-readable format.

=cut
sub display_genes {
    my ($genes) = @_;
    for my $gene (@$genes) {
        print $gene->to_string(), "\n";
    }
}

# --- Main Program ---
if (@ARGV) {
    my $filename = $ARGV[0];
    my $genes = parse_genome_file($filename);
    display_genes($genes);
    my $json_file = "$filename-byPL.json";
    export_to_json($genes, $json_file) if $filename =~ /\.gen$/;
    print "Exported to $json_file\n";
}

=head1 EXAMPLES

=head2 Example: Parsing a Genome File
perl creatures1genomeAnalysis.pl dad1.gen

=head2 Example Output
Gene 001: Type=0x02 (Creature), Subtype=0x01 (Genus), SwitchOn=0 (Embryo), SexDep=None, Mutability=None
Exported to dad1.gen.json

=cut

File: biosilico-creatures/src/main/resources/creatures/articlesAroundAndScripts/scripts/creatures1genomeAnalysis.py
#!/usr/bin/env python3
"""
creatures1genomeAnalysis.py - Parse and analyze Creatures genome files.

Usage:
    python3 creatures1genomeAnalysis.py <filename.gen> [--json <output.json>] [--csv <output.csv>]

Description:
    This script parses Creatures genome files (`.gen`) and extracts gene information.
    It supports all gene types and subtypes, including Brain Lobes, Biochemistry Reactions, and Creature Instincts.
    The output includes human-readable descriptions and JSON/CSV exports.

Gene Structure:
    - Header: 8 bytes (gene marker, type, subtype, etc.)
    - Data: Variable length, parsed based on gene type/subtype.
"""

import struct
import json
import csv
from dataclasses import dataclass, field
from typing import List, Dict, Any, Optional

# --- Reference Data Sources ---
# Source: creatures1BrainMapCells_GenesHeader.txt
LOBE_NAMES = {
    0: "Perception",
    1: "Drive",
    2: "Stimulus Source",
    3: "Verb",
    4: "Noun",
    5: "General Sense",
    6: "Decision",
    7: "Attention",
    8: "Concept",
    9: "Regulator",  # Creatures 2+
}

# Source: creaturesDevelopmentRessources.pdf, p.8-12
CHEM_NAMES = {
    0x00: "Pain",
    0x01: "Need for Pleasure",
    0x02: "Hunger",
    0x03: "Coldness",
    0x04: "Hotness",
    0x0A: "Fear",
    0x0B: "Boredom",
    0x1E: "Reward",
    0x1F: "Punishment",
}

# Source: creatures1BrainMapCells_GenesHeader.txt, p.21
ACTION_NAMES = {
    0: "Quiescent",
    1: "Push (Activate 1)",
    2: "Pull (Activate 2)",
    3: "Stop (Deactivate)",
    4: "Come (Approach)",
    5: "Run (Retreat)",
    6: "Get",
    7: "Drop",
    8: "Think/Say",
    9: "Sleep/Rest",
    10: "Left",
    11: "Right",
}

TYPE_NAMES = {
    0x00: "Brain",
    0x01: "Biochemistry",
    0x02: "Creature",
}

SUBTYPE_NAMES = {
    0x00: {0x00: "Lobe"},
    0x01: {
        0x00: "Receptor",
        0x01: "Emitter",
        0x02: "Reaction",
        0x03: "Half-Lives",
        0x04: "Initial Concentration",
    },
    0x02: {
        0x00: "Stimulus",
        0x01: "Genus",
        0x02: "Appearance",
        0x03: "Pose",
        0x04: "Gait",
        0x05: "Instinct",
        0x06: "Pigment",
        0x07: "Pigment Bleed",
    },
}

STAGE_NAMES = ["Embryo", "Child", "Youth", "Adolescent", "Adult", "Senior", "Old"]

@dataclass
class Gene:
    """Represents a Creatures gene with parsed data."""
    type: int
    subtype: int
    number: int
    switch_on: int
    sex_dep: str
    mutability: str
    data: List[int]
    parsed_data: Dict[str, Any] = field(default_factory=dict)

    def parse_reaction(self) -> None:
        """Parses Reaction genes (Biochemistry type)."""
        if not (self.type == 0x01 and self.subtype == 0x02):
            return
        if not self.data or len(self.data) < 4:
            return

        reactants = []
        products = []
        for i in range(0, min(4, len(self.data)), 2):
            if i + 1 >= len(self.data):
                break
            chem = self.data[i + 1]
            proportion = self.data[i]
            container = reactants if i < 2 else products
            chem_name = CHEM_NAMES.get(chem, f"Chem{chem:02X}")
            container.append({
                "chem": chem,
                "proportion": proportion,
                "chem_name": chem_name,
            })

        rate = self.data[8] if len(self.data) >= 9 else 1
        self.parsed_data = {
            "reactants": reactants,
            "products": products,
            "rate": rate,
        }

    def parse_instinct(self) -> None:
        """Parses Instinct genes (Creature type)."""
        if not (self.type == 0x02 and self.subtype == 0x05):
            return
        if not self.data or len(self.data) < 9:
            return

        lobe1, cell1, lobe2, cell2, lobe3, cell3, action, reward_punish, amount = self.data[:9]
        self.parsed_data = {
            "conditions": [
                {"lobe": LOBE_NAMES.get(lobe1, f"Lobe{lobe1:02X}"), "cell": cell1},
                {"lobe": LOBE_NAMES.get(lobe2, f"Lobe{lobe2:02X}"), "cell": cell2},
                {"lobe": LOBE_NAMES.get(lobe3, f"Lobe{lobe3:02X}"), "cell": cell3},
            ],
            "action": ACTION_NAMES.get(action, f"Action{action:02X}"),
            "reward_punish": "Reward" if reward_punish == 0 else "Punish",
            "amount": amount,
        }

    def parse_lobe(self) -> None:
        """Parses Brain Lobe genes."""
        if not (self.type == 0x00 and self.subtype == 0x00):
            return
        if not self.data or len(self.data) < 5:
            return

        x, y, width, height, perception_link = self.data[:5]
        self.parsed_data = {
            "position": {"x": x, "y": y},
            "size": {"width": width, "height": height},
            "perception_link": "Yes" if perception_link else "No",
            "neurons": width * height,
        }
    # TODO add for emitter, receptor, initconc

    def __str__(self) -> str:
        """Returns a human-readable string representation of the gene."""
        base_str = (
            f"Gene {self.number:03d}: Type=0x{self.type:02X} ({TYPE_NAMES.get(self.type, 'Unknown')}), "
            f"Subtype=0x{self.subtype:02X} ({SUBTYPE_NAMES.get(self.type, {}).get(self.subtype, 'Unknown')}), "
            f"SwitchOn={self.switch_on} ({STAGE_NAMES[self.switch_on] if self.switch_on < len(STAGE_NAMES) else 'Unknown'}), "
            f"SexDep={self.sex_dep}, Mutability={self.mutability}"
        )

        if self.type == 0x01 and self.subtype == 0x02 and "reactants" in self.parsed_data:
            reaction = self.parsed_data
            base_str += (
                f"\n  Reaction: {' + '.join(f'{r['proportion']} {r['chem_name']}' for r in reaction['reactants'])} → "
                f"{' + '.join(f'{p['proportion']} {p['chem_name']}' for p in reaction['products'])} "
                f"(Rate: {reaction['rate']})"
            )
        elif self.type == 0x02 and self.subtype == 0x05 and "conditions" in self.parsed_data:
            instinct = self.parsed_data
            base_str += (
                f"\n  Instinct: IF {' AND '.join(f'{c['lobe']}[{c['cell']}]' for c in instinct['conditions'])} "
                f"THEN {instinct['action']} ({instinct['reward_punish']}: {instinct['amount']})"
            )
        elif self.type == 0x00 and self.subtype == 0x00 and "position" in self.parsed_data:
            lobe = self.parsed_data
            base_str += (
                f"\n  Lobe: {LOBE_NAMES.get(self.number, f'Lobe{self.number:02X}')} "
                f"at ({lobe['position']['x']},{lobe['position']['y']}), "
                f"Size: {lobe['size']['width']}x{lobe['size']['height']}, "
                f"Neurons: {lobe['neurons']}, Perception Link: {lobe['perception_link']}"
            )

        return base_str

def parse_genome_file(filename: str) -> List[Gene]:
    """
    Parses a .gen file and returns a list of Gene objects.

    Args:
        filename: Path to the .gen file.

    Returns:
        List of Gene objects.
    """
    with open(filename, "rb") as f:
        content = f.read()

    genes = []
    offset = 0
    while offset < len(content):
        header = content[offset:offset+8]
        if len(header) < 7:  # 4s (4) + 3x B (3) = 7 bytes minimum
            offset += 1
            continue

        gene_marker = header[:4]
        if gene_marker != b'gene':
            offset += 1
            continue

        type_, subtype, number = struct.unpack("B B B", header[4:7])
        switch_on = content[offset + 7] if offset + 7 < len(content) else 0
        sex_mut = content[offset + 8] if offset + 8 < len(content) else 0

        sex_dep = "Female" if sex_mut & 0x10 else ("Male" if sex_mut & 0x08 else "None")
        mutability = []
        if sex_mut & 0x01: mutability.append("Mutable")
        if sex_mut & 0x02: mutability.append("Duplicable")
        if sex_mut & 0x04: mutability.append("Deletable")
        mutability_str = ",".join(mutability) or "None"

        data_length = content[offset + 9] if offset + 9 < len(content) else 0
        data = []
        if data_length > 0 and offset + 10 + data_length <= len(content):
            data = list(struct.unpack(f"{data_length}B", content[offset + 10:offset + 10 + data_length]))

        gene = Gene(
            type=type_,
            subtype=subtype,
            number=number,
            switch_on=switch_on,
            sex_dep=sex_dep,
            mutability=mutability_str,
            data=data,
        )

        gene.parse_reaction()
        gene.parse_instinct()
        gene.parse_lobe()
        # TODO add for emitter, receptor, initconc

        genes.append(gene)
        offset += 10 + data_length

    return genes

def export_to_json(genes: List[Gene], filename: str) -> None:
    """Exports gene data to a JSON file."""
    json_genes = []
    for gene in genes:
        json_gene = {
            "type": f"0x{gene.type:02X}",
            "type_name": TYPE_NAMES.get(gene.type, "Unknown"),
            "subtype": f"0x{gene.subtype:02X}",
            "subtype_name": SUBTYPE_NAMES.get(gene.type, {}).get(gene.subtype, "Unknown"),
            "number": gene.number,
            "switch_on": gene.switch_on,
            "switch_on_name": STAGE_NAMES[gene.switch_on] if gene.switch_on < len(STAGE_NAMES) else "Unknown",
            "sex_dep": gene.sex_dep,
            "mutability": gene.mutability,
            "data": gene.data,
        }
        if gene.parsed_data:
            json_gene["parsed_data"] = gene.parsed_data
        json_genes.append(json_gene)

    with open(filename, "w") as f:
        json.dump(json_genes, f, indent=2)

def export_to_csv(genes: List[Gene], filename: str) -> None:
    """Exports gene data to a CSV file."""
    with open(filename, "w", newline='') as f:
        writer = csv.writer(f)
        writer.writerow([
            "Number", "Type", "Subtype", "SwitchOn", "SexDep", "Mutability",
            "Data", "ParsedData"
        ])
        for gene in genes:
            writer.writerow([
                gene.number,
                f"0x{gene.type:02X} ({TYPE_NAMES.get(gene.type, 'Unknown')})",
                f"0x{gene.subtype:02X} ({SUBTYPE_NAMES.get(gene.type, {}).get(gene.subtype, 'Unknown')})",
                f"{gene.switch_on} ({STAGE_NAMES[gene.switch_on] if gene.switch_on < len(STAGE_NAMES) else 'Unknown'})",
                gene.sex_dep,
                gene.mutability,
                " ".join(map(str, gene.data)),
                str(gene.parsed_data) if gene.parsed_data else "",
            ])

def display_genes(genes: List[Gene]) -> None:
    """Displays gene information in a human-readable format."""
    for gene in genes:
        print(gene)

if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser(description="Parse Creatures genome files.")
    parser.add_argument("file", help="Genome file to parse")
    parser.add_argument("--json", help="Export to JSON file")
    parser.add_argument("--csv", help="Export to CSV file")
    args = parser.parse_args()

    genes = parse_genome_file(args.file)
    display_genes(genes)

    if args.json:
        export_to_json(genes, args.json)
    if args.csv:
        export_to_csv(genes, args.csv)
